lib/Basic/Subprocess.cpp

```diff
@@ -436,17 +452,47 @@ void llbuild::basic::spawnProcess(
 
     // If we have been cancelled since we started, do nothing.
     if (!wasCancelled) {
-      int result = posix_spawn(&pid, args[0], /*file_actions=*/&fileActions,
-                               /*attrp=*/&attributes, const_cast<char**>(args.data()),
-                               const_cast<char* const*>(environment.getEnvp()));
+      int result = 0;
+
+#ifdef __APPLE__
+      int oldCWD = -1;
+      int chdirErr = 0;
+      if (attr.workingDir.size()) {
+        oldCWD = open(".", O_RDONLY);
+        if (pthread_fchdir_np(-1) == -1 && oldCWD != -1) {
+          // no existing per-thread cwd
+          close(oldCWD); oldCWD = -1;
+        }
```

Comment:
I think this is unnecessary -- we own the thread, so we should be able to keep track directly of what the latest per-thread WD was. I'd like it if we could avoid any extraneous sys calls in the common case where almost all commands will run in the same directory.